runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPipelineFuser.java

```diff
@@ -0,0 +1,321 @@
+      rootConsumers.addAll(getRootEnvTransforms(pTransformNode));
+    }
+    fusePipeline(groupSiblings(rootConsumers));
+  }
+
+  public static FusedPipeline fuse(Pipeline p) {
+    GreedyPipelineFuser fuser = new GreedyPipelineFuser(p);
+    return FusedPipeline.of(fuser.stages, fuser.unfusedTransforms);
+  }
+
+  /**
+   * Fuses a {@link Pipeline} into a collection of {@link ExecutableStage}.
+   *
+   * <p>The input is the initial collection of siblings sets which will be fused into {@link
+   * ExecutableStage stages}. A sibling in this context represents a pair of (PCollection,
+   * PTransform), where the PTransform consumes input elements on a per-element basis from the
+   * PCollection, represented by a {@link CollectionConsumer}. A sibling set is a collection of
+   * siblings which can execute within a single {@link ExecutableStage}, determined by {@link
+   * GreedyPCollectionFusers#isCompatible(PTransformNode, PTransformNode, QueryablePipeline)}.
+   *
+   * <p>While a pending sibling set exists:
+   *
+   * <ul>
+   *   <li>Retrieve a pending sibling set from the front of the queue.
+   *   <li>If the pending sibling set has already been created, continue. Each materialized {@link
+   *       PTransformNode} can be consumed by any number of {@link ExecutableStage stages}, but each
+   *       {@link PTransformNode} may only be present in a single stage rooted at a single {@link
+   *       PCollectionNode}, otherwise it will process elements of that {@link PCollectionNode}
+   *       multiple times.
+   *   <li>Create a {@link GreedilyFusedExecutableStage} with those siblings as the initial
+   *       consuming transforms of the stage
+   *   <li>For each materialized {@link PCollectionNode}, find all of the descendant in-environment
+   *       consumers. See {@link #getDescendantConsumersInEnv(PCollectionNode)} for details.
+   *   <li>Construct all of the sibling sets from the descendant in-environment consumers, and add
+   *       them to the queue of sibling sets.
+   * </ul>
+   */
+  private void fusePipeline(NavigableSet<NavigableSet<CollectionConsumer>> initialConsumers) {
+    Queue<Set<CollectionConsumer>> pendingSiblingSets = new ArrayDeque<>();
+    pendingSiblingSets.addAll(initialConsumers);
+    while (!pendingSiblingSets.isEmpty()) {
+      // Only introduce new PCollection consumers. Not performing this introduces potential
+      // duplicate paths through the pipeline.
+      Set<CollectionConsumer> candidateSiblings = pendingSiblingSets.poll();
+      Set<CollectionConsumer> siblingSet =
+          Sets.difference(candidateSiblings, consumedCollectionsAndTransforms.keySet());
+      checkState(
+          siblingSet.equals(candidateSiblings) || siblingSet.isEmpty(),
+          "Inconsistent collection of siblings reported for a %s. Initial attempt missed %s",
+          PCollectionNode.class.getSimpleName(),
+          siblingSet);
+      if (siblingSet.isEmpty()) {
```

Comment:
This is one of the two cases that `checkState` permits - we've materialized a `PCollection` in multiple stages, and we've encountered the duplicate here.